apps/api/src/app/events/usecases/map-trigger-recipients/map-trigger-recipients.use-case.ts

```diff
@@ -0,0 +1,139 @@
+  TriggerRecipients,
+  TriggerRecipientsPayload,
+} from '@novu/node';
+import {
+  EnvironmentId,
+  LogCodeEnum,
+  LogStatusEnum,
+  OrganizationId,
+  TopicId,
+  TriggerRecipientsTypeEnum,
+  UserId,
+} from '@novu/shared';
+
+import { MapTriggerRecipientsCommand } from './map-trigger-recipients.command';
+
+import { CreateLog, CreateLogCommand } from '../../../logs/usecases/create-log';
+import { GetTopicSubscribersCommand, GetTopicSubscribersUseCase } from '../../../topics/use-cases';
+
+interface ILogTopicSubscribersPayload {
+  environmentId: EnvironmentId;
+  organizationId: OrganizationId;
+  topicId: TopicId;
+  transactionId: string;
+  userId: UserId;
+}
+
+@Injectable()
+export class MapTriggerRecipients {
+  constructor(private createLog: CreateLog, private getTopicSubscribers: GetTopicSubscribersUseCase) {}
+
+  async execute(command: MapTriggerRecipientsCommand): Promise<ISubscribersDefine[]> {
+    const { environmentId, organizationId, recipients, transactionId, userId } = command;
+
+    const mappedRecipients = Array.isArray(recipients) ? recipients : [recipients];
+
+    const simpleSubscribers: ISubscribersDefine[] = this.findSubscribers(mappedRecipients);
+
+    const topicSubscribers: ISubscribersDefine[] = await this.getSubscribersFromAllTopics(
+      transactionId,
+      environmentId,
+      organizationId,
+      userId,
+      mappedRecipients
+    );
+
+    return [...simpleSubscribers, ...topicSubscribers];
+  }
+
+  private async getSubscribersFromAllTopics(
+    transactionId: string,
+    environmentId: EnvironmentId,
+    organizationId: OrganizationId,
+    userId: UserId,
+    recipients: TriggerRecipients
+  ): Promise<ISubscribersDefine[]> {
+    const topics = this.findTopics(recipients);
+
+    const topicSubscribers: ISubscribersDefine[] = [];
+
+    for (const topic of topics) {
+      try {
+        const getTopicSubscribersCommand = GetTopicSubscribersCommand.create({
+          environmentId,
+          topicId: topic?.topicId || '',
+          organizationId,
+          userId,
+        });
+        const response = await this.getTopicSubscribers.execute(getTopicSubscribersCommand);
+        const { subscribers } = response;
+
+        subscribers.forEach((subscriber) => topicSubscribers.push({ subscriberId: subscriber }));
+      } catch (error) {
+        this.logTopicSubscribersError({
+          environmentId,
+          organizationId,
+          topicId: topic?.topicId || '',
+          transactionId,
+          userId,
+        });
+      }
+    }
+
+    return topicSubscribers;
+  }
+
+  public mapSubscriber(subscriber: TriggerRecipientSubscriber): ISubscribersDefine {
+    if (typeof subscriber === 'string') {
+      return { subscriberId: subscriber };
+    }
+
+    return subscriber;
+  }
+
+  private findSubscribers(recipients: TriggerRecipients): ISubscribersDefine[] {
+    const isNotTopic = (recipient: TriggerRecipientSubscriber): recipient is TriggerRecipientSubscriber =>
```

Comment:
maybe move `isNotTopic` outside of this function?

Reply:
I usually have this functions on top but this one was created just because the guard to stop TS complaining. Otherwise I wouldn't have created. Do you think it will be used in other places therefore is a good idea to move it out?